### ISSUE TITLE
Unit testing framework for KeystoneJS projects

### ISSUE THREAD

**Issue by JedWatson** [keystonejs/keystone#216]
It's not clear how to write unit tests for a KeystoneJS project.

We should create a mini-framework for testing that initialises Keystone then allows tests to be run (which could be written using any of the available frameworks, e.g. mocha)

**Comment by captainchung** [comment]
Feels like a bad hack, but does anyone who is looking to unit test have any thoughts to this gist?
https://gist.github.com/captainchung/620429d314a849ef470e

**Comment by moll** [comment]
@captainchung: Start Keystone in an async `beforeEach` block and wait till it starts. Timeouts are fragile.
Other than that, nothing extra required indeed.

**Comment by captainchung** [comment]
Thanks @moll. Within the `beforeEach`, how do I know when keystone has started / to call done()?

**Comment by moll** [comment]
Keystone.start or something took a callback IIRC.

**Comment by captainchung** [comment]
@moll i just updated my gist. is this what you were thinking with Keystone.start? https://gist.github.com/captainchung/620429d314a849ef470e

**Comment by moll** [comment]
Basically, but I believe you can simplify all this by dozen lines by just doing something like
```javascript
before(Keystone.start.bind(Keystone))
```

Last I remember, though, Keystone wasn't too happy about being started and stopped for every test. Using a `before` in the outermost scope, as opposed to `beforeEach`, might work around that by starting Keystone only once per whole test suite.

**Comment by webteckie** [comment]
For what I do with keystone, my integration testing works fine with restarting keystone on Jasmine's beforeEach.  For unit testing, I'm using https://www.npmjs.com/package/keystonejs-stub, however, it relies heavily on mocking.  Currently I'm stuck on correctly mocking mongoose virtuals.  @trystant added some logic for virtuals but it doesn't work for me; thus, I'm looking to update it by also simply being able to mock virtuals but so far last time I tackled it it proved to be a challenge and put it on hold.

**Comment by cameronjroe** [comment]
I'd highly recommend [karma](https://github.com/karma-runner/karma). It's a great test runner that allows for both browser e2e testing and unit testing. I'd be happy to start a PR to help implement this. Here's a few plugins that might be of interest:

- https://github.com/karma-runner/karma-mocha
- https://github.com/xdissent/karma-chai
- https://github.com/nikku/karma-browserify
- https://github.com/caitp/karma-coveralls

**Comment by foobar8675** [comment]
if there was a sample for unit testing using karma for keystone, maybe as
part of the example project, i think that would be super helpful.

On Tue, Sep 1, 2015 at 6:59 AM, Cameron J Roe <notifications@github.com>
wrote:

> I'd highly recommend karma <https://github.com/karma-runner/karma>. It's
> a great test runner that allows for both browser e2e testing and unit
> testing. I'd be happy to start a PR to help implement this. Here's a few
> plugins that might be of interest:
>
>    - https://github.com/karma-runner/karma-mocha
>    - https://github.com/xdissent/karma-chai
>    - https://github.com/nikku/karma-browserify
>    - https://github.com/caitp/karma-coveralls
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/keystonejs/keystone/issues/216#issuecomment-136731444>
> .
>



-- 
Regards,
Matt

**Comment by cameronjroe** [comment]
Ah, I actually meant to add this internally to keystone as well. Maybe there's another issue for that.

**Comment by creynders** [comment]
[Added this to the wiki](https://github.com/keystonejs/keystone/wiki/Knowledge-Base) since there's some valuable information here.
